operators/pkg/controller/elasticsearch/driver/downscale.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/elastic/cloud-on-k8s/operators/pkg/utils/k8s"
+	appsv1 "k8s.io/api/apps/v1"
+)
+
+func (d *defaultDriver) HandleDownscale(
+	expectedStatefulSets sset.StatefulSetList,
+	actualStatefulSets sset.StatefulSetList,
+	esClient esclient.Client,
+	observedState observer.State,
+	reconcileState *reconcile.State,
+) *reconciler.Results {
+	results := &reconciler.Results{}
+
+	// compute the list of nodes leaving the cluster, from which
+	// data should be migrated away
+	leavingNodes := []string{}
+
+	// process each statefulset for downscale
+	for i, actual := range actualStatefulSets {
+		expected, shouldExist := expectedStatefulSets.GetByName(actual.Name)
+		targetReplicas := int32(0) // sset removal
+		if shouldExist {           // sset downscale
+			targetReplicas = sset.Replicas(expected)
+		}
+		leaving, removalResult := d.scaleStatefulSetDown(actualStatefulSets, &actualStatefulSets[i], targetReplicas, esClient, observedState, reconcileState)
+		results.WithResults(removalResult)
+		if removalResult.HasError() {
+			return results
+		}
+		leavingNodes = append(leavingNodes, leaving...)
+	}
+
+	// migrate data away from nodes leaving the cluster
+	log.V(1).Info("Migrating data away from nodes", "nodes", leavingNodes)
+	if err := migration.MigrateData(esClient, leavingNodes); err != nil {
+		return results.WithError(err)
+	}
+
+	return results
+}
+
+func (d *defaultDriver) scaleStatefulSetDown(
+	allStatefulSets sset.StatefulSetList,
+	ssetToScaleDown *appsv1.StatefulSet,
+	targetReplicas int32,
+	esClient esclient.Client,
+	observedState observer.State,
+	reconcileState *reconcile.State,
+) ([]string, *reconciler.Results) {
+	results := &reconciler.Results{}
+	logger := log.WithValues("statefulset", k8s.ExtractNamespacedName(ssetToScaleDown))
+
+	if sset.Replicas(*ssetToScaleDown) == 0 && targetReplicas == 0 {
+		// no replicas expected, StatefulSet can be safely deleted
+		logger.Info("Deleting statefulset", "namespace", ssetToScaleDown.Namespace, "name", ssetToScaleDown.Name)
+		if err := d.Client.Delete(ssetToScaleDown); err != nil {
+			return nil, results.WithError(err)
+		}
+	}
+	// copy the current replicas, to be decremented with nodes to remove
+	initialReplicas := sset.Replicas(*ssetToScaleDown)
+	updatedReplicas := initialReplicas
+
+	// leaving nodes names can be built from StatefulSet name and ordinals
+	// nodes are ordered by highest ordinal first
+	var leavingNodes []string
+	for i := initialReplicas - 1; i > targetReplicas-1; i-- {
+		leavingNodes = append(leavingNodes, sset.PodName(ssetToScaleDown.Name, i))
+	}
+
+	// TODO: don't remove last master/last data nodes?
+	// TODO: detect cases where data migration cannot happen since no nodes to host shards?
+
+	for _, node := range leavingNodes {
+		if migration.IsMigratingData(observedState, node, leavingNodes) {
+			// data migration not over yet: schedule a requeue
+			logger.V(1).Info("Data migration not over yet, skipping node deletion", "node", node)
```

Comment:
Do we want to set the status to migrating here? I'm not sure what the right way to propagate that up is. This PR is already pretty large but I wonder if that's something we should do relatively soon after.

Reply:
Good catch, it was there before. Adding it back.